### PR TITLE
Feature/cobra cli

### LINKED ISSUE #75: One executable for all CLI commands by using Cobra
# Proposal for a single executable based Scicat CLI tool
I would like to create a PR for a code rewrite that would put all of the separate executables defined in `cmd/` into one, where each component would get their own subcommand.

This would basically mean: Instead of calling the command this way:
```
./[command_name] (flags) (params)
```
We'd have the following convention:
```
./scicat-cli [command_name] (flags) (params)
```
This would be achieved through the [Cobra library](https://github.com/spf13/cobra/).

As an example of this concept, here's my [globus library](https://github.com/SwissOpenEM/globus) with a built-in cli implementation.

- [x] #79 
- [x] configure CI to allow for prereleases
- [ ] v3 -> main merging

### PR DESCRIPTION
**Pull request by consolethinks**
## Issue
Resolves #75 

## PR Description
It copies the pre-existing code in each executable's main function, adapts them to Cobra's way of getting flags and parameters. 

Each command supported by this one executable should behave the same way, for a given set of flags and parameters, as the corresponding tool that it was based on.

## Notes
This PR should be tested against PSI usecases still.

### CODE REVIEW

**Comment by sbliven** [comment]
This introduces as `cmd/cmd` folder. This is a bit confusing. Should we rename the module?

**Comment by consolethinks** [comment]
This last commit should address the double cmd folder structure by renaming the inner cmd folder to "commands" (as it contains command implementations of the CLI tool). If need be, the outer cmd folder could be renamed to cli as well, but I think the former is the "standard" in Go.

File: cmd/commands/datasetArchiver.go
Comment (minottic) [inline cmd/commands/datasetArchiver.go]: was this block removed by mistake?

File: cmd/main.go
Comment (minottic) [inline cmd/main.go:5]: could we test here that when we call this main with an arg (e.g. datasetIngestor) the datasetIngestor from `commands` is executed?

File: cmd/commands/datasetArchiver.go
Comment (minottic) [reply cmd/commands/datasetArchiver.go]: having this will allow to check that passing variables by reference did not introduce any error

File: cmd/commands/datasetPublishData.go
Comment (minottic) [inline cmd/commands/datasetPublishData.go:103]: why do these need to become internal functions? is this because the command is now a pointer? I am not a fun of internal functions, as they will be very hard to test later

File: cmd/commands/datasetRetriever.go
Comment (minottic) [inline cmd/commands/datasetRetriever.go:60]: see comment above

**Review by minottic — COMMENTED** [review]
I feel tests are missing and, once ready, I will be worried to merge all the main changes as a single commit. Also the CI breaks. 

This is a very useful PR, thanks!

File: cmd/commands/root.go
Comment (consolethinks) [reply cmd/commands/root.go:15]: I might add a default action since the version parameter is common between all commands

File: cmd/commands/datasetArchiver.go
Comment (consolethinks) [reply cmd/commands/datasetArchiver.go]: it was deleted because the code was original only copied from before I added tests, this is readded in the latest commit.

File: cmd/main.go
Comment (consolethinks) [reply cmd/main.go:5]: I have done something similar to that in the latest commit (calling Execute() instead of main, but it does effectively the same thing)

File: cmd/commands/datasetPublishData.go
Comment (consolethinks) [reply cmd/commands/datasetPublishData.go:103]: they are internal because the functions are only used once and only exist to make the code more readable. If they weren't internal, the functions' definitions would be shared among all the commands.  This wasn't an issue when we had lots of little "main" packages, but now there's a bigger "cmd" package that ties everything together.

**Comment by sbliven** [comment]
Not to hold up this merge even more, but are we ok with removing the old tools? I had imagined keeping the old executable around for backwards compatibility. After removing them we will have to maintain two branches until we drop support for the old cli names.

**Review by minottic — APPROVED** [review]

**Comment by minottic** [comment]
I think we can double check the CI and the version comparison and use this eventually.
image
This will allow us to get rid of the v3 branch and merge everything into main, even before officially releasing. 

From memory, the release workflow in the CI overrides the "prerelease" and "latest" flags, so we should check that

**Pull request merged**
